File: sdk/dart-holons/test/holonrpc_http_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:holons/holons.dart';
import 'package:test/test.dart';

void main() {
  group('holon-rpc http', () {
    test('invoke uses unary POST and decodes JSON-RPC result', () async {
      final server = await _startHTTPServer((server) {
        server.register('echo.v1.Echo/Ping', (params) async => params);
      });
      addTearDown(() async {
        await server.close();
      });

      final rawClient = HttpClient();
      addTearDown(() {
        rawClient.close(force: true);
      });

      final request = await rawClient
          .postUrl(Uri.parse('${server.address}/echo.v1.Echo/Ping'));
      request.headers.contentType = ContentType.json;
      request.headers.set(HttpHeaders.acceptHeader, 'application/json');
      request.write('{"message":"hello"}');

      final response = await request.close();
      expect(response.statusCode, equals(HttpStatus.ok));
      expect(
        response.headers.contentType?.mimeType,
        equals('application/json'),
      );

      final payload = jsonDecode(
        await utf8.decoder.bind(response).join(),
      ) as Map<String, dynamic>;
      expect(
        (payload['result'] as Map<String, dynamic>)['message'],
        equals('hello'),
      );

      final client = HolonRPCHTTPClient(server.address);
      addTearDown(client.close);

      final result = await client.invoke(
        'echo.v1.Echo/Ping',
        params: const <String, dynamic>{'message': 'hola'},
      );
      expect(result['message'], equals('hola'));
    });

    test('stream uses POST SSE', () async {
      final server = await _startHTTPServer((server) {
        server.registerStream('build.v1.Build/Watch', (params, send) async {
          expect(params['project'], equals('myapp'));
          await send(<String, dynamic>{
            'status': 'building',
            'progress': 42,
          });
          await send(<String, dynamic>{
            'status': 'done',
            'progress': 100,
          });
        });
      });
      addTearDown(() async {
        await server.close();
      });

      final client = HolonRPCHTTPClient(server.address);
      addTearDown(client.close);

      final events = await client.stream(
        'build.v1.Build/Watch',
        params: const <String, dynamic>{'project': 'myapp'},
      );

      expect(events, hasLength(3));
      expect(events[0].event, equals('message'));
      expect(events[0].id, equals('1'));
      expect(events[0].result['status'], equals('building'));
      expect(events[1].result['status'], equals('done'));
      expect(events[2].event, equals('done'));
    });

    test('streamQuery uses GET SSE', () async {
      final server = await _startHTTPServer((server) {
        server.registerStream('build.v1.Build/Watch', (params, send) async {
          expect(params['project'], equals('myapp'));
          await send(<String, dynamic>{'status': 'watching'});
        });
      });
      addTearDown(() async {
        await server.close();
      });

      final client = HolonRPCHTTPClient(server.address);
      addTearDown(client.close);

      final events = await client.streamQuery(
        'build.v1.Build/Watch',
        params: const <String, String>{'project': 'myapp'},
      );

      expect(events, hasLength(2));
      expect(events[0].result['status'], equals('watching'));
      expect(events[1].event, equals('done'));
    });

    test('server answers CORS preflight', () async {
      final server = await _startHTTPServer(null);
      addTearDown(() async {
        await server.close();
      });

      final client = HttpClient();
      addTearDown(() {
        client.close(force: true);
      });

      final request = await client.openUrl(
          'OPTIONS', Uri.parse('${server.address}/echo.v1.Echo/Ping'));
      request.headers.set('origin', 'https://example.test');

      final response = await request.close();
      expect(response.statusCode, equals(HttpStatus.noContent));
      expect(
        response.headers.value('Access-Control-Allow-Origin'),
        equals('https://example.test'),
      );
      expect(
        response.headers.value('Access-Control-Allow-Methods'),
        equals('GET, POST, OPTIONS'),
      );
    });

    test('method not found returns code 5', () async {
      final server = await _startHTTPServer(null);
      addTearDown(() async {
        await server.close();
      });

      final client = HolonRPCHTTPClient(server.address);
      addTearDown(client.close);

      await expectLater(
        () => client.invoke('missing.v1.Service/Method'),
        throwsA(
          isA<HolonRPCResponseException>()
              .having((error) => error.code, 'code', equals(5)),
        ),
      );
    });
  });
}

Future<HolonRPCHTTPServer> _startHTTPServer(
  void Function(HolonRPCHTTPServer server)? register,
) async {
  final server = HolonRPCHTTPServer('http://127.0.0.1:0/api/v1/rpc');
  register?.call(server);
  await server.start();
  return server;
}

File: sdk/node-holons/tests/holonrpc_http.test.js
'use strict';

const { describe, it } = require('node:test');
const assert = require('node:assert/strict');
const http = require('node:http');

const { holonrpc } = require('../src/index');

function startHTTPRPCServer(t) {
    let nextID = 1;

    const server = http.createServer(async (req, res) => {
        setCORSHeaders(req, res);

        if (req.method === 'OPTIONS') {
            res.writeHead(204);
            res.end();
            return;
        }

        const url = new URL(req.url, 'http://127.0.0.1');
        const method = url.pathname.replace(/^\/api\/v1\/rpc\//, '');

        if (req.method === 'POST' && method === 'echo.v1.Echo/Ping') {
            const params = await readJSON(req);
            writeJSON(res, 200, {
                jsonrpc: '2.0',
                id: `h${nextID++}`,
                result: {
                    message: String(params.message || ''),
                    sdk: 'rest+sse-server',
                },
            });
            return;
        }

        if (acceptsSSE(req) && method === 'build.v1.Build/Watch') {
            res.writeHead(200, {
                'Content-Type': 'text/event-stream',
                'Cache-Control': 'no-cache',
                Connection: 'keep-alive',
            });

            if (req.method === 'POST') {
                const params = await readJSON(req);
                assert.equal(params.project, 'myapp');
                writeSSE(res, 'message', '1', {
                    jsonrpc: '2.0',
                    id: 'h1',
                    result: { status: 'building', progress: 42 },
                });
                writeSSE(res, 'message', '2', {
                    jsonrpc: '2.0',
                    id: 'h1',
                    result: { status: 'done', progress: 100 },
                });
                writeDone(res);
                return;
            }

            if (req.method === 'GET') {
                assert.equal(url.searchParams.get('project'), 'myapp');
                writeSSE(res, 'message', '1', {
                    jsonrpc: '2.0',
                    id: 'h2',
                    result: { status: 'watching' },
                });
                writeDone(res);
                return;
            }
        }

        writeJSON(res, 404, {
            jsonrpc: '2.0',
            id: 'h0',
            error: {
                code: 5,
                message: `method "${method}" not found`,
            },
        });
    });

    return new Promise((resolve, reject) => {
        server.once('error', reject);
        server.listen(0, '127.0.0.1', () => {
            const address = server.address();
            const baseURL = `http://127.0.0.1:${address.port}/api/v1/rpc`;

            t.after(async () => {
                await new Promise((done) => server.close(() => done()));
            });

            resolve(baseURL);
        });
    });
}

function setCORSHeaders(req, res) {
    const origin = req.headers.origin || '*';
    res.setHeader('Access-Control-Allow-Origin', origin);
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Accept, Last-Event-ID');
    res.setHeader('Access-Control-Max-Age', '86400');
    res.setHeader('Vary', 'Origin');
}

function acceptsSSE(req) {
    return String(req.headers.accept || '').toLowerCase().includes('text/event-stream');
}

function readJSON(req) {
    return new Promise((resolve, reject) => {
        const chunks = [];
        req.on('data', (chunk) => chunks.push(Buffer.from(chunk)));
        req.on('end', () => {
            const text = Buffer.concat(chunks).toString('utf8').trim();
            resolve(text ? JSON.parse(text) : {});
        });
        req.on('error', reject);
    });
}

function writeJSON(res, statusCode, payload) {
    res.writeHead(statusCode, {
        'Content-Type': 'application/json',
    });
    res.end(JSON.stringify(payload));
}

function writeSSE(res, event, id, payload) {
    res.write(`event: ${event}\n`);
    res.write(`id: ${id}\n`);
    res.write(`data: ${JSON.stringify(payload)}\n\n`);
}

function writeDone(res) {
    res.write('event: done\n');
    res.write('data:\n\n');
    res.end();
}

describe('holonrpc http+sse', () => {
    it('invokes unary methods over http://', async (t) => {
        const baseURL = await startHTTPRPCServer(t);
        const client = new holonrpc.HolonRPCClient();
        t.after(async () => client.close());

        await client.connect(baseURL);
        assert.equal(client.connected(), true);

        const result = await client.invoke('echo.v1.Echo/Ping', { message: 'http-audit' });
        assert.equal(result.message, 'http-audit');
        assert.equal(result.sdk, 'rest+sse-server');
    });

    it('accepts rest+sse:// URLs as an http:// alias', async (t) => {
        const baseURL = await startHTTPRPCServer(t);
        const client = new holonrpc.HolonRPCClient();
        t.after(async () => client.close());

        await client.connect(baseURL.replace('http://', 'rest+sse://'));
        const result = await client.invoke('echo.v1.Echo/Ping', { message: 'alias-audit' });
        assert.equal(result.message, 'alias-audit');
    });

    it('streams POST SSE responses', async (t) => {
        const baseURL = await startHTTPRPCServer(t);
        const client = new holonrpc.HolonRPCClient();
        t.after(async () => client.close());

        await client.connect(baseURL);
        const events = await client.stream('build.v1.Build/Watch', { project: 'myapp' });

        assert.equal(events.length, 3);
        assert.equal(events[0].event, 'message');
        assert.equal(events[0].id, '1');
        assert.equal(events[0].result.status, 'building');
        assert.equal(events[1].result.status, 'done');
        assert.equal(events[2].event, 'done');
    });

    it('streams GET SSE responses', async (t) => {
        const baseURL = await startHTTPRPCServer(t);
        const client = new holonrpc.HolonRPCClient();
        t.after(async () => client.close());

        await client.connect(baseURL);
        const events = await client.streamQuery('build.v1.Build/Watch', { project: 'myapp' });

        assert.equal(events.length, 2);
        assert.equal(events[0].event, 'message');
        assert.equal(events[0].result.status, 'watching');
        assert.equal(events[1].event, 'done');
    });
});
